src/System.Runtime.Extensions/src/System/IO/StreamWriter.cs

```diff
@@ -334,26 +334,45 @@ public override void Write(char[] buffer)
 
             CheckAsyncTaskInProgress();
 
-            int index = 0;
             int count = buffer.Length;
-            while (count > 0)
+			
+            // Threshold of 4 was chosen after running perf tests
+            if (count <= 4)
             {
-                if (_charPos == _charLen)
+                for (int i = 0; i < count; i++)
```

Comment:
Looks like it does eliminate range checks via loop cloning which isn't that great because it increases code size. It's best to use `i < array.Length` whenever possible, I already cleaned up a few such cases in corefx.

It's probably best to just write
```C#
if (buffer.Length < 4) {
    for (...; i < buffer.Length; ...) ...
}
else {
    int count = buffer.Length;
    ...
}
```